### PR TITLE
fixing edge cases in the serializer

### PR DESCRIPTION
**Pull request by agroszer**

### CODE REVIEW

File: src/pjpersist/serialize.py [hunk 1]
```diff
@@ -547,6 +549,7 @@ def get_non_persistent_object(self, state, obj):
 
         # this method must NOT change the passed in state dict
         state = dict(state)
+        setState = True
```
Comment (kedder) [inline src/pjpersist/serialize.py]: Why do you need this variable?

File: src/pjpersist/serialize.py [hunk 2]
```diff
@@ -356,15 +356,17 @@ def get_state(self, obj, pobj=None):
...
             for key, value in obj.items():
                 data.append((key, self.get_state(value, pobj)))
-                has_non_string_key |= not isinstance(key, six.string_types)
-                if (not isinstance(key, six.string_types) or '\0' in key):
-                    has_non_string_key = True
-            if not has_non_string_key:
-                # The easy case: all keys are strings:
+                if (not isinstance(key, six.string_types) or  # non-string
+                    '\0' in key or  # string having a 0x00
+                    key=='dict_data'  # a key with our special marker
```
Comment (kedder) [inline src/pjpersist/serialize.py]: Huh? what is `\0`? what is `dict_data`? What if there is `0x1` character in the string? What if there is `0x2` character? What if  there is "list_data" in it?

Comment (agroszer) [reply src/pjpersist/serialize.py]: `\0` is a chr(0) looks like it's mongo legacy, with PG/JSON as investigation shows you just cannot store a chr(0) be it in a key or just a simple string

Checking for `dict_data` is there to avoid failing in `get_object` as this is a special marker when encoding non-string keys.
See `doctest_ObjectWriter_get_state_mappings` "Another edge case is when the input has a `dict_data` key"

Comment (agroszer) [reply src/pjpersist/serialize.py]: We can't really change this `dict_data` encoding, because then we would need a migration or BBB code.

File: src/pjpersist/serialize.py (hunk 1 above)
Comment (agroszer) [reply src/pjpersist/serialize.py]: because L571 will figure whether it needs to set state, the other cases always want to set state

either this or I could add a private method to set state...

Comment (kedder) [reply src/pjpersist/serialize.py]: Isn't the purpose of this MR is to always set state? When we don't want to do that?

File: src/pjpersist/serialize.py (hunk 2 above)
Comment (kedder) [reply src/pjpersist/serialize.py]: I don't understand why special marker is called `dict_data`. Shouldn't it be called `special_marker` instead?

File: src/pjpersist/serialize.py (hunk 1 above)
Comment (agroszer) [reply src/pjpersist/serialize.py]: We don't want to set state when `state` has `_py_factory` and `_py_factory_args` but nothing else. That would 'clean out' the object, resetting state that was set by `_py_factory_args`

File: src/pjpersist/serialize.py (hunk 2 above)
Comment (agroszer) [reply src/pjpersist/serialize.py]: Yes that's why it's there, but do we want migrations? Look at the recent push, I made it a global.

File: src/pjpersist/serialize.py (hunk 1 above)
Comment (kedder) [reply src/pjpersist/serialize.py]: I don't get it.  Why would it "clean out" the object? What does it have to do with `_py_factory_args`? If empty `state` somehow overrides `_py_factory_args`, wouldn't non-empty state also override it? Why `_py_factory_args` is there at all, if all state can be stored in `state` itself?

File: src/pjpersist/serialize.py [hunk 3]
```diff
@@ -66,6 +66,9 @@
     '__builtin__.set': set,
 }
 
+DICT_NON_STRING_KEY_MARKER = 'dict_data'
```
Comment (kedder) [inline src/pjpersist/serialize.py]: Let's call it `'non_string_key_marker'` instead. `dict_data` doesn't make any sense, if it is not dictionary data.

File: src/pjpersist/tests/test_serialize.py [hunk 4]
```diff
@@ -421,6 +421,15 @@ def doctest_ObjectWriter_get_state_mappings():
...
       >>> pprint.pprint(writer.get_state({1: 'one', 2: 'two', 3: 'three'}))
       {'dict_data': [(1, 'one'), (2, 'two'), (3, 'three')]}
+
+    Another edge case is when the input has a `dict_data` key
+
+      >>> writer = serialize.ObjectWriter(None)
+      >>> state = writer.get_state(
+      ...     {'1': 1, 'dict_data': 'works?'})
+      >>> pprint.pprint(sorted(state['dict_data']))
+      [('1', 1), ('dict_data', 'works?')]
```
Comment (kedder) [inline src/pjpersist/tests/test_serialize.py]: Do we need to test an edge case where there is a `"list_data"` in the input?

File: src/pjpersist/serialize.py (hunk 2 above)
Comment (kedder) [reply src/pjpersist/serialize.py]: Why are you talking about migrations?

File: src/pjpersist/serialize.py (hunk 1 above)
Comment (agroszer) [reply src/pjpersist/serialize.py]: Example:
```
state={
    '_py_factory': 'PersistentDict',
    '_py_factory_args': {some_dict_items}
}

            factory = self.simple_resolve(state.pop('_py_factory'))
            factory_args = self.get_object(state.pop('_py_factory_args'), obj)
            sub_obj = factory(*factory_args)
```
now `sub_obj` has `some_dict_items`, `state` is `{}`

if we let the code run into
```
sub_obj_state = self.get_object(state, obj)
sub_obj.__setstate__(sub_obj_state)
```

`sub_obj` will get `some_dict_items` wiped by the 'empty' state

pretty obscure edge case, `doctest_ObjectReader_get_non_persistent_object_py_factory` failed on that

File: src/pjpersist/serialize.py (hunk 2 above)
Comment (agroszer) [reply src/pjpersist/serialize.py]: Because that weirdo `dict_data` gets persisted to the database.
See `doctest_ObjectWriter_get_state_mappings` how state looks.

File: src/pjpersist/serialize.py (hunk 1 above)
Comment (kedder) [reply src/pjpersist/serialize.py]: Wait, so `sub_obj` will wipe itself? Isn't it the problem of `sub_obj`, and not the pjpersist?

File: src/pjpersist/tests/test_serialize.py (hunk 4 above)
Comment (agroszer) [reply src/pjpersist/tests/test_serialize.py]: don't think so, we don't add `list_data`

File: src/pjpersist/serialize.py (hunk 2 above)
Comment (kedder) [reply src/pjpersist/serialize.py]: Well, `list_data` will also persist itself if it is part of the object state. Should we also handle it here?

File: src/pjpersist/serialize.py (hunk 3 above)
Comment (agroszer) [reply src/pjpersist/serialize.py]: This edge case is dict only. JSON(B) accepts only strings as keys.

File: src/pjpersist/tests/test_serialize.py (hunk 4 above)
Comment (kedder) [reply src/pjpersist/tests/test_serialize.py]: Why not?

File: src/pjpersist/serialize.py (hunk 3 above)
Comment (kedder) [reply src/pjpersist/serialize.py]: What does this have to do with `dict_data`?
